fix: localImportMap not works

chilingling opened the pull request:
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Background and solution
<!--
1. Describe the problem and the scenario.
2. New features need to be described and attached with renderings.
3. Screenshots or GIFs involving UI/Interaction changes/Bugfix before and after modification are required.
-->

### What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

### What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced local CDN resource localization with configurable import maps and copy rules for improved offline and restricted network support.
  - Added a Vite plugin for copying and transforming files with import path rewriting during build.
  - Enabled dynamic generation of import map URLs and centralized CDN domain, version, and local bundle path configuration.
  - Added environment variables and example configs to customize local CDN paths and behavior.
  - Enhanced preview and canvas components to support local CDN resource loading.

- **Bug Fixes**
  - Improved handling and filtering of invalid or unmatched CDN URLs during build processing.

- **Documentation**
  - Added detailed solution guides and documentation entries for local CDN usage, configuration, and implementation.

- **Tests**
  - Added comprehensive automated tests validating local CDN features, including dependency copying and bundle localization.

- **Chores**
  - Updated dev dependencies and scripts to include testing frameworks and static copy plugins.
  - Refactored import map and bundle metadata to support destructuring imports and dynamic CDN URL construction.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chilingling committed:
fix: localImportMap not works

chilingling committed:
fix: remove irrelevant code

chilingling committed:
fix: change config layer

chilingling committed:
fix: add test case

rhlin commented on designer-demo/env/.env.alpha:
```diff
@@ -1,12 +1,11 @@
-VITE_LOCAL_IMPORT_MAPS=false
-VITE_LOCAL_BUNDLE_DEPS=false
+VITE_LOCAL_IMPORT_MAPS=true
+VITE_LOCAL_BUNDLE_DEPS=true
```
可以考虑默认保留false

hexqi replied:
默认值不要修改

gene9831 commented on docs/catalog.json:
```diff
@@ -78,7 +78,8 @@
             { "title": "设计器中引入第三方组件库", "name": "third-party-library-in-designer.md" },
-            { "title": "物料同步方案", "name": "material-sync-solution.md" }
+            { "title": "物料同步方案", "name": "material-sync-solution.md" },
+            { "title": "本地化CDN方案", "name": "local-cdn.md" }
```
可以执行下当前目录下的 `generateCatalogMd.js` 来更新 `docs/README.md`

chilingling replied:
已执行

rhlin commented on docs/solutions/local-cdn.md:
```diff
@@ -0,0 +1,207 @@
+    filePathInPackage: '/'
+  },
+  '@opentiny/vue-runtime': {
+    filePathInPackage: '/dist3/'
```
这个配置不能解释 @opentiny/vue映射到vue-runtime ？
我觉得这个dist3是无感的才对， 所以好像不应该存在这个配置文件

chilingling replied:
这里默认配置是复制策略的默认配置，这几个都需要复制整个文件夹，所以作为了默认的配置。

rhlin commented on packages/build/vite-config/src/localCdnFile/copyBundleDeps.js:
```diff
@@ -13,29 +13,30 @@ import {
-  bundle.data?.materials?.components?.forEach((component) => {
-    if (component.npm) {
-      const possibleUrl = [component.npm.script, component.npm.css]
+  bundle.data?.materials?.packages?.forEach((packageItem) => {
```
原油的component协议里的scripts和css得并没有废弃，是否要保留支持？

chilingling replied:
已修改，同时兼容

rhlin commented on packages/build/vite-config/src/localCdnFile/copyBundleDeps.js:
```diff
@@ -50,9 +51,10 @@ export function copyBundleDeps({
-  )
+  const cdnFiles = extraBundleCdnLink(bundleFile, originCdnPrefix)
+    .map((url) => getCdnPathNpmInfoForSingleFile(url, originCdnPrefix, base, dir, false, bundleTempDir))
+    .filter(Boolean)
```
最好是写语义化一点的？ 这个路径是可能为null是不？ 那如果是null了是如何处理？

chilingling replied:
已增加注释说明

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,386 @@
+
+      config.define['import.meta.env.VITE_CDN_DOMAIN'] = JSON.stringify(`./${cdnDir}`)
+      // 使用本地 CDN 时，强制设置CDN类型为 local
+      config.define['import.meta.env.VITE_CDN_TYPE'] = JSON.stringify('local')
```
这里的强制修改env 可能引起麻烦，  理论上不应该回头去覆盖文件变量，  通过vite_xxx_xxx  然后回设置同一个文件的vite_xxx，算是引起一种耦合和数据流问题， 下游不应该重定义上游

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,386 @@
+ * @param {string} filename - 文件名
+ * @returns {string} - 处理后的内容
+ */
+function transform(content, filename) {
```
不应该使用通用的名字代表特殊的内容处理，需要重命名函数

chilingling replied:
已修改成 `replaceJsImportPaths` 相对语义化的命名

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,386 @@
+    // 使用自定义的copyPlugin替代直接调用copy
+    copyPlugin(targetFiles)
+  ]
+}
```
这个文件似乎有很多不同层次的插件， 逻辑分析的 文件系统操作的，请考虑拆成每个插件一个js文件，把不同层次的业务内容摘出去，比如默认的importmap文件要从最外层使用的地方配置默认文件

chilingling replied:
已提取 `cdnCopyPlugin` 和 `createEnvReplacementPlugin` 两个 vite 插件到 `vite-plugins` 文件夹下

rhlin commented on packages/build/vite-config/src/localCdnFile/locateCdnNpmInfo.js:
```diff
@@ -48,9 +48,30 @@ export function getCdnPathNpmInfoForSingleFile(
+  }
+
+  // 使用匹配到的结果
+  const match = npmmirrorMatch || unpkgMatch
```
不建议内置这么具体选择的逻辑

chilingling replied:
合并两个正则表达式后已不存在

rhlin commented on packages/build/vite-config/src/localCdnFile/locateCdnNpmInfo.js:
```diff
@@ -111,9 +132,22 @@ export function getCdnPathNpmInfoForPackage(
     new RegExp(`^${originCdnPrefix}/?(?<packageName>.+?)@(?<versionDemand>[^/]+)(?<filePathInPackage>.*?)$`)
-  ).groups
+  )
+  const npmmirrorMatch = url.match(
```
这两个正则应该可以合一的

chilingling replied:
已合并处理

rhlin commented on packages/common/js/importMap/import-map.json:
```diff
@@ -1,8 +1,8 @@
 {
   "imports": {
-    "vue": "${VITE_CDN_DOMAIN}/vue${versionDelimiter}3.4.23${fileDelimiter}/dist/vue.runtime.esm-browser.js",
+    "vue": "${VITE_CDN_DOMAIN}/vue${versionDelimiter}3.4.23${fileDelimiter}/dist/vue.runtime.esm-browser.prod.js",
```
当时是画布和预览使用开发模式下的vue方便调试问题（获得更多的warning警告信息）

chilingling replied:
已修改成  `vue.runtime.esm-browser.js`

chilingling committed:
fix: add base path to local cdn path

chilingling committed:
fix: merge unpkg regexp and npmmirror regexp

chilingling committed:
fix: generate docs catalog

chilingling committed:
fix: add comment for url filter logic

chilingling committed:
refactor: optimize function naming in local CDN plugin

rename transform function to replaceJsImportPaths to improve code readability and semantics. The new name more clearly expresses the actual functionality of the function: replacing import paths in JS files.

chilingling committed:
refactor: optimize localCDN plugin implementation.

提取独立的插件 cdnCopyPlugin、createEnvReplacementPlugin 到 vite-plugin 文件夹中。

chilingling committed:
fix: support legacy npm potocol in CDN link extraction

chilingling committed:
fix: optimize local CDN plugin implementation

hexqi commented on packages/design-core/vite.config.js:
```diff
@@ -62,7 +62,7 @@ export default defineConfig({
         }
       },
-      external: ['vue', 'monaco-editor', 'prettier', /@opentiny\/vue.*/, '@opentiny/tiny-engine-meta-register']
+      external: ['vue', 'monaco-editor', 'prettier', /@opentiny\/vue.*/, /^@opentiny\/tiny-engine*/]
```
这个PR里为什么要把@opentiny/tiny-engine*依赖都排除呢

rhlin replied:
不再二次打包吧

chilingling replied:
因为 common 中使用了虚拟模块 `virtual:import-map`，在 design-core 打包的时候，无法提前得到真正的 import-map 模块内容。所以需要做一个 external。

rhlin replied:
这里可能有个错漏 应该是 `e.*`

hexqi commented on .gitignore:
```diff
@@ -6,6 +6,7 @@ yarn.lock
 pnpm-lock.yaml
 lerna-debug.log
 packages/design-core/bundle-deps
+designer-demo/bundle-deps
```
![image](https://github.com/user-attachments/assets/a7ff9a44-f567-41f1-819e-a7fa81e973ec)
updateTemplate脚本可以更新下

rhlin commented on packages/build/vite-config/src/localCdnFile/copyBundleDeps.js:
```diff
@@ -13,29 +13,42 @@ import {
 export function replaceBundleCdnLink(bundle, fileMap) {
-  bundle.data?.materials?.components?.forEach((component) => {
-    if (component.npm) {
+  bundle.data?.materials?.packages?.forEach((packageItem) => {
```
`extraBundleCdnLink`有兼容老的协议，到了`replaceBundleCdnLInk`的时候缺乏兼容了

chilingling replied:
已兼容

rhlin commented on packages/build/vite-config/src/localCdnFile/copyBundleDeps.js:
```diff
@@ -50,9 +63,12 @@ export function copyBundleDeps({
-  )
+  const cdnFiles = extraBundleCdnLink(bundleFile, originCdnPrefix)
+    .map((url) => getCdnPathNpmInfoForSingleFile(url, originCdnPrefix, base, dir, false, bundleTempDir))
+    // 比如 url 前缀跟 originCdnPrefix 不匹配，或者 url 格式不正确 的场景有可能会导致匹配失败
```
理论上不会出现前缀跟originCdnPrefix不匹配场景，可以再斟酌一下

chilingling replied:
用户有可能内部打包的物料和官方物料混合在一起：

```javascript
{
"packages": [
  {
    "name": "自定打包的组件库",
    "package": "美好出行-xxxxx-component",
    "version": "3.20.0",
    "destructuring": true,
    "script": "https://192.168.1.10/xxxx.mjs",
    "css": "https://192.168.1.10/xxxx.css"
  },
  {
    "name": "TinyVue组件库",
    "package": "@opentiny/vue",
    "version": "3.20.0",
    "destructuring": true,
    "script": "https://registry.npmmirror.com/@opentiny/vue-runtime/~3.20/files/dist3/tiny-vue-pc.mjs",
    "css": "https://registry.npmmirror.com/@opentiny/vue-theme/~3.20/files/index.css"
  },
  {
    "name": "element-plus组件库",
    "package": "element-plus",
    "version": "2.4.2",
    "script": "https://registry.npmmirror.com/element-plus/2.4.2/files/dist/index.full.mjs",
    "css": "https://registry.npmmirror.com/element-plus/2.4.2/files/dist/index.css"
  }
]
}
```

rhlin replied:
copyBundleDeps.js L34已经判断果是要替换的前缀的了？

chilingling replied:
```javascript
// copyBundleDeps.js extraBundleCdnLink
// 这里要判断的是前缀匹配。
if (url?.startsWith(originCdnPrefix)) {
```

前缀匹配之后，还有从整个 url 里面根据特定的正则匹配出来版本号、路径，如果匹配不成功，也就是无法提取出来有效的版本、路径，那么就无法做本地化，也需要过滤掉：

```javascript
// 生成复制单个文件所需要的信息
export function getCdnPathNpmInfoForSingleFile(
  url, // cdn托管的npm文件地址数组
  originCdnPrefix, // cdn的前缀
  base, // build构建的base（BASE_URL）参数
  dir, // 复制到目标的文件目录
  transformIContent = false, // 是否需要转换内容, 如果传入url实际为目录则不会转会
  tempDir = 'bundle-deps' // 新安装包的安装目录
) {
  const baseSlash = base.endsWith('/') ? '' : '/'
  const match = extractPackageInfo(url, originCdnPrefix)

  // ！！！看这里
  if (!match) {
    return null
  }
```

rhlin commented on packages/build/vite-config/src/localCdnFile/import-map.js:
```diff
@@ -0,0 +1,33 @@
+export const importMapConfig = {
```
这个文件暂时不明白为啥在build vite插件里面 这个应该属于业务内容，应该通过传参进来，与插件要解耦

chilingling replied:
背景：
- 画布（@opentiny/tiny-engine-canvas） 需要 import-map
- 页面预览 (preview) 需要 import-map
- 上述两个地方依赖的 import-map  大致重合（vue、vue-i18n、物料后续会逐步解耦出去）

我们需要 import-map 本地化：所以我们拿到将上述两个地方的 import-map 依赖，将依赖提前打包到产物，然后让原来的 import-map 拿到本地化之后的域名

原来的方案：
- 将 import-map.json 放置到 common，让画布和页面预览依赖读取。
- import-map 的域名 `${VITE_CDN_DOMAIN}` 保持字符串。
- 本地化打包之后，改写环境变量 `${VITE_CDN_DOMAIN}`，在运行时将 `${VITE_CDN_DOMAIN}` 字符串替换成环境变量的值。（就是插件包不感知是不是 localCDN，他们只关心环境变量的值）

新方案：
改动 commit: [`e9b75c4` (#1242)](https://github.com/opentiny/tiny-engine/pull/1242/commits/e9b75c49697187eaffa02f0ddc07e2898605ff0b)

改动原因：https://github.com/opentiny/tiny-engine/pull/1242#discussion_r2015744519

![截屏2025-04-23 21 38 11](https://github.com/user-attachments/assets/ac25619e-1f72-4b68-b267-690495b277df)

方案改动点：
- 如果不能改环境变量 `${VITE_CDN_DOMAIN}` 的值，那么我们需要将真正的 localCDN 相对路径写入到真正的模块中。

**使用虚拟模块的原因**
如果不使用虚拟模块，无法在二开工程打包的时候，将真正的 localCDN 相对路径写入到真正的模块中。

**使用虚拟模块后**
如果开启了 CDN 本地化，则将替换后的内容该写成模块内容，再参与真正的打包。
如果没有开启 CDN 本地化，则将原样的内容写入模块内容，参与真正的打包。

如果有更好的方案，欢迎提出来。

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,204 @@
+
+const logger = console
+
+// 默认的复制配置，这几个 package 需要复制整个目录，所以需要有默认配置进行复制整个目录
```
这个配置应该从应用层传参数进来 vite插件不要内置业务会比较好

chilingling replied:
当前 tiny-vue 物料还是内置的，需要解耦 tiny-vue 物料之后，再将此部分内置逻辑解耦出来。

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,204 @@
+ * @param {string} versionTarget - 目标版本号,来源于 package.json 的 version, 不能包含 ^ 或 ~ 开头
+ * @returns {boolean} - 是否相同
+ */
+const compareIsSameVersion = (versionOrigin, versionTarget) => {
```
此处属于简单实现， 可以抽取单个文件为semver版本号字符串的相等compare，作为utils单独一个文件， 或者更建议看看有无三方库直接使用，把逻辑关注点分离

chilingling replied:
这里的相等比较其实不是严格的相等比较。
这里是为了避免重复下载依赖包做的比较，比如我的 cdn 指定版本为：`^3.20.0`
那么，只要我的 `node_module` 里面的 package.json 的版本是大版本号相同就可以。
如果指定版本为 `~3.20.0`，那么只要我的 `node_module` 里面的 package.json 的版本是 `3.20.x` 就可以认为相同。
所以我理解不是严格比较的通用逻辑，后续如果其他插件或者业务也有类似的逻辑，可以抽取成一个通用的 `utils`。

rhlin commented on packages/build/vite-config/src/localCdnFile/localCdnPlugin.js:
```diff
@@ -0,0 +1,204 @@
+    return versionOrigin.slice(1, 3) === versionTarget.slice(0, 3)
+  }
+
+  return false
```
compare函数需要返回 0 1 或者-1
is函数返回true false，
需要修改一下函数名明确

chilingling replied:
这里没懂为什么 `compare` 函数需要返回 0、1 或者 -1，是哪里的规范吗？

rhlin replied:
约定俗成

chilingling replied:
哪里的约定俗成？有相关链接吗？

rhlin commented on packages/build/vite-config/src/default-config.js:
```diff
@@ -160,22 +161,33 @@ export function useTinyEngineBaseConfig(engineConfig) {
+    }
+  }
+
+  config.plugins.push(
```
已经再包一层插件的话，插件应该一体化， L183-L88应该放到 168的if分支块内，甚至应该放到localCdnPlugin内

chilingling replied:
这里不能放置到  `168` 行的 if 分支条件块内部。理由详见：https://github.com/opentiny/tiny-engine/pull/1242#discussion_r2056110000

因为如果不是 localImportMap 了，也需映射虚拟模块。

rhlin commented on packages/build/vite-config/src/vite-plugins/createDynamicImportMapPlugin.js:
```diff
@@ -0,0 +1,37 @@
+          return `export default ${newContent}`
+        }
+
+        return `export default ${JSON.stringify(importMapConfig)}`
```
这里的两个条件分支的处理层级有点不一样， 待进一步分析

chilingling replied:
处理的层级不一样是什么意思？这个插件是决定终态的虚拟模块的内容的。

rhlin commented on packages/canvas/DesignCanvas/src/importMap.ts:
```diff
@@ -1,38 +1,48 @@
 
+  if (customImportMap?.imports?.[pkgName]) {
+    return customImportMap.imports[pkgName]
+      .replace('${VITE_CDN_DOMAIN}', VITE_CDN_DOMAIN)
```
考虑通过vite插件同一个层级一起处理了？

chilingling replied:
这里实际分为两种场景：

1. localCDN 的场景：
由于 https://github.com/opentiny/tiny-engine/pull/1242#discussion_r2015744519 的 意见提出，localCDN 的场景会将实际的 `${VITE_CDN_DOMAIN}` 在构建态实际写入。使得 `VITE_CDN_DOMAIN`  环境变量失效。

2. 非 localCDN 的场景：

当前仍然保留了 `VITE_CDN_DOMAIN` 等环境变量的运行时读取和替换

如果 localCDN 和非 localCDN 的场景都将 import-map 的  CDN 等变量直接在构建态解决的话，我理解就不需要配置到 `.env` 环境变量了。应该作为构建配置项存在。

rhlin commented on packages/canvas/DesignCanvas/src/importMap.ts:
```diff
@@ -1,38 +1,48 @@
-      `${VITE_CDN_DOMAIN}/@opentiny/vue-theme${versionDelimiter}${importMapVersions.tinyVue}${fileDelimiter}/index.css`,
-      `${VITE_CDN_DOMAIN}/element-plus${versionDelimiter}2.4.2${fileDelimiter}/dist/index.css`
-    ]
+    importStyles: []
```
style理论上应该和imports 同个时间清理，应该是一对一绑定的

rhlin commented on packages/common/vite.config.ts:
```diff
@@ -66,7 +66,8 @@ export default defineConfig({
         /^prettier.*/,
-        /^@babel.*/
+        /^@babel.*/,
+        /^virtual:import-map$/
```
这里可能得约定一个其他名字比如约定成包名会更优雅一定

rhlin replied:
可能是由于该文件不应该分这这个包？

chilingling replied:
是否需要分包，可以参考 https://github.com/opentiny/tiny-engine/pull/1242#discussion_r2056110000 的背景讨论

rhlin commented on packages/design-core/src/preview/src/preview/importMap.js:
```diff
@@ -12,7 +12,7 @@
 import { useEnv } from '@opentiny/tiny-engine-meta-register'
 import { getSearchParams } from './http'
-import importMapJSON from './importMap.json'
+import { importMapConfig as importMapJSON } from '@opentiny/tiny-engine-common/js/importMap/index'
```
这个路径过于具体（有耦合倾向，需要维持api稳定），common包理论上 只暴露到js目录这一层

rhlin commented on packages/build/vite-config/src/default-config.js:
```diff
@@ -160,22 +161,33 @@ export function useTinyEngineBaseConfig(engineConfig) {
   )
 
+  // 添加本地化CDN插件支持
+  if (isLocalImportMap) {
```
如果要改变写法， 需要把isCopyBundleDeps一起改了，保持工整对称

rhlin commented on packages/build/vite-config/src/default-config.js:
```diff
@@ -160,22 +161,33 @@ export function useTinyEngineBaseConfig(engineConfig) {
+    const logger = console
+    logger.log('[local-cdn-plugin]: Initializing local CDN plugin')
+
+    const cdnPlugins = localCdnPlugin({
```
这个插件似乎只是处理preview预览？ 因为上面的copyBundleDeps是处理画布物料的， 需要考虑合适的名字

chilingling committed:
fix: replace bundle cdnLink support old protocol

chilingling committed:
feat: optimize local cdn implementation

chilingling committed:
fix: del useless test case

chilingling committed:
doc: update localCDN docs

chilingling committed:
feat: use semver to compare version

chilingling committed:
fix: only process .js or .mjs file content

chilingling commented on designer-demo/tests/localCdnBasic.test.js:
```diff
@@ -0,0 +1,88 @@
+  afterAll(() => {
+    restoreEnvFile(envAlphaPath)
+  })
+  
```
good idea, i will consider it on the next move.

chilingling committed:
fix: making some variables more semantic.

chilingling committed:
fix: change by review comment

chilingling committed:
fix: adjust some code catalog.

chilingling committed:
fix: support importmap styles

chilingling committed:
fix: del duplicated config

chilingling committed:
fix: update updateTemplate

chilingling committed:
fix: update doc file name

hexqi commented on docs/solutions/import-map-local.md:
```diff
@@ -0,0 +1,212 @@
+# 本地化 import-map CDN方案
```
Github Copilot优化后的内容：

```markdown
下面是优化后的文档内容，调整了语句表达和结构，使其更通顺、更易理解：

---

# 本地化 Import-Map CDN 方案

## 概述

TinyEngine 在画布和预览中使用了 import-map 的方式来加载依赖，例如 `vue`、`vue-i18n` 和物料等。这些 import-map 默认依赖于 npmmirror 的 CDN 服务。然而，在某些企业场景中，由于可用性和稳定性要求，无法依赖外部的 CDN 服务。

为了解决这一问题，可以采用以下方案：
- 搭建企业私有网络中的 unpkg 服务
- 采用本文档介绍的 Import-Map CDN 本地化方案

本地化 Import-Map CDN 的核心思想是：在构建过程中，将 import-map 所需的远程 CDN 资源替换为构建产物中的本地文件。其主要优点包括：
1. 减少对外部 CDN 的依赖，提升应用可靠性。
2. 支持离线或内网环境中对 import-map 所需资源的访问。
3. 加快资源加载速度，提升应用性能。

## 使用方法

### 启用 Import-Map CDN 本地化

按照以下步骤启用 Import-Map CDN 本地化功能：

1. **修改环境变量**

在 `.env.alpha`、`.env.production` 或 `.env.development` 文件中添加以下配置：

```bash
# 启用 import-map 的本地化功能
VITE_LOCAL_IMPORT_MAPS=true

# 启用物料的本地化功能（需确保物料包可通过 npm 下载）
VITE_LOCAL_BUNDLE_DEPS=true

# 定义本地 CDN 资源存放目录，默认为 local-cdn-static
VITE_LOCAL_BUNDLE_PATH=local-cdn-static
```

2. **（可选）在 `vite.config.js` 中传入自定义配置**

可通过 `importMapLocalConfig` 选项配置导入映射规则和文件复制行为：

```javascript
const baseConfig = useTinyEngineBaseConfig({
  importMapLocalConfig: {
    importMap: { imports: { ... } },
    copy: { ... }
  }
  // 其他配置...
})
```

### 配置选项

本地化 CDN 支持以下配置选项：

| 参数                      | 类型   | 默认值                              | 说明                          |
|---------------------------|--------|-------------------------------------|-------------------------------|
| `importMapLocalConfig`    | Object | `{ importMap: { imports: {} }, copy: {} }` | 本地化 CDN 配置对象              |
| `importMapLocalConfig.importMap` | Object | `{ imports: {} }`                  | 导入映射配置，定义需要本地化的 CDN 依赖 |
| `importMapLocalConfig.copy`      | Object | `{}`                               | 自定义文件复制配置，可覆盖默认行为   |

#### 导入映射（importMap）说明

`importMapLocalConfig.importMap` 定义了需要本地化的 CDN 依赖。示例格式如下：

```json
{
  "imports": {
    "vue": "${VITE_CDN_DOMAIN}/vue${versionDelimiter}3.2.37${fileDelimiter}/dist/vue.global.prod.js",
    "vue-router": "${VITE_CDN_DOMAIN}/vue-router${versionDelimiter}4.1.5${fileDelimiter}/dist/vue-router.global.js",
    "pinia": "${VITE_CDN_DOMAIN}/pinia${versionDelimiter}2.0.23${fileDelimiter}/dist/pinia.iife.prod.js"
  }
}
```

**占位符说明：**
- `${VITE_CDN_DOMAIN}`：CDN 域名占位符，将被替换为本地路径。
- `${versionDelimiter}`：版本分隔符，用于分隔包名和版本。
- `${fileDelimiter}`：文件路径分隔符，用于分隔版本和文件路径。

#### 文件复制（copy）说明

`importMapLocalConfig.copy` 可自定义特定包的文件复制配置。默认配置如下：

```javascript
{
  '@opentiny/vue-theme': { filePathInPackage: '/' },
  '@opentiny/vue-renderless': { filePathInPackage: '/' },
  '@opentiny/vue-runtime': { filePathInPackage: '/dist3/' },
  '@vue/devtools-api': { filePathInPackage: '/' }
}
```

支持以下选项：
- `filePathInPackage`：指定复制的包内路径，复制整个包时设置为 `/`。
- `version`：可选，覆盖包的版本号（只影响下载的版本号，不影响实际请求 URL 的版本号）。

示例自定义配置：

```javascript
{
  'vue': { filePathInPackage: '/dist/' },
  'element-plus': { filePathInPackage: '/dist/index.css', version: '2.2.0' }
}
```

### 处理 Bundle 文件中的 CDN 链接

在 `.env` 文件中启用 `VITE_LOCAL_BUNDLE_DEPS=true`，处理物料文件 `bundle.json` 中的 CDN 链接。

示例：
```bash
VITE_CDN_DOMAIN=https://unpkg.com
```

如果 `bundle.json` 文件中的 URL 前缀与 `VITE_CDN_DOMAIN` 匹配，则会解析为本地路径。

---

## 实现原理

Import-Map CDN 本地化的主要实现步骤如下：
1. **解析导入映射**：分析 `importMapLocalConfig` 和默认配置，识别需要本地化的依赖。
2. **依赖安装**：检查和安装所需依赖包。
3. **文件复制与转换**：将资源复制到目标目录，并替换路径为本地化链接。
4. **处理 Bundle 文件**：替换 bundle 文件中的 CDN 链接，生成新的文件。

## 注意事项

1. 本地化 CDN 会增加构建输出大小，但显著提高应用可靠性和性能。
2. 某些特殊格式的 CDN URL 可能需要额外配置。

--- 

你可以根据优化后的内容直接替换原文档。

```

hexqi commented on docs/solutions/import-map-local.md:
```diff
@@ -0,0 +1,212 @@
+
+`importMapLocalConfig.importMap` 是一个包含 `imports` 属性的对象，它定义了需要本地化的CDN依赖。在插件内部，它会与默认的导入映射配置合并。
+
+import-map.json 的格式示例：
```
建议改成importMap格式示例，这里没用import-map.json

hexqi commented on docs/solutions/import-map-local.md:
````diff
@@ -0,0 +1,212 @@
+```json
+{
+  "imports": {
+    "vue": "${VITE_CDN_DOMAIN}/vue${versionDelimiter}3.2.37${fileDelimiter}/dist/vue.global.prod.js",
````
可以加下这个路径是怎么得来的，还有补充下面的三个变量一般是什么内容

hexqi commented on docs/solutions/import-map-local.md:
```diff
@@ -0,0 +1,212 @@
+
+## 使用方法
+
+### 修改环境变量使用 import-map CDN 本地化方案
```
我们需要统一下术语，import-map是技术相关不容易理解， 建议就叫“CDN依赖本地化“方案

chilingling committed:
docs: optimize local-cdn docs

chilingling committed:
fix: update bundle variable
